### DISCUSSION TITLE
Régulation en dehors des plages de température min

### DISCUSSION THREAD

**Discussion by dliouville** [jmcollin78/versatile_thermostat#579] [Q&A]
Bonjour,

Déjà merci beaucoup pour tout le travail effectué.

Je soulève une petite question, éventuellement un bug je ne sais pas encore, voici la description de la situation : 
- j'ai un chauffage par pompe a chaleur air-air, on ne peut pas définir une température inférieure à 16°C sur les appareils.
- j'ai configuré un thermostat pour gérer les consoles de chauffage et définis une plage de température avec un minimum de 16°C car la consigne ne peut pas descendre en dessous.
- Dans la configuration du thermostat j'ai demandé une gestion de régulation "forte" car quand il fait froid (en dessous de 6°C extérieur environ) le maintient de la température n'est pas assez efficace dans les autres modes.
- J'ai configuré une programmation pour gérer la température en fonction d'un planning sur la journée.

Le comportement observé : 
Au moment de mon planning de baisse de température de chauffe, qui passe de 19°C à 17°C, si ma température dans la pièce est comme aujourd'hui à une 20aine de °C et qu'il n'y a pas besoin de chauffe, le thermostat envoie une consigne de température à 12°C ce qui est inapplicable par mes appareils (16°C min) et la consigne boucle toutes les quelques minutes. A la limite ça ne serait pas dérangeant si ça n'était que du code qui tourne pour rien ... mais ... hélas ... mes équipements émettent un BIP très désagréable à chaque demande de changement de consigne (et ce BIP sonore n'est pas désactivable, demandes remontées sans succès au constructeur (viessman)), du coup je me retrouve avec des BIPs en boucle .....

Donc au final, il me semble qu'il manquerai un garde-fou dans le code pour empêcher une consigne de régulation de descendre en dessous de la température minimale permise dans la configuration.

J'ai un peu regardé le code, je suppose qu'il y a 2 solutions : 
- soit au niveau de _send_regulated_temperature
- soit dans calculate_regulated_temperature
Pour définir que la température ne puisse être inférieure à CONF_TEMP_MIN

Cordialement,
David Liouville

**Comment by jmcollin78** [comment]
Bonjour @dliouville ,

C'est normalement déjà le cas. Toutes les températures régulées envoyées sont bornées dans les valeurs min/max. Je suis très étonné de ton retour.
```
@overrides
    def cap_sent_value(self, value) -> float:
        """Try to adapt the target temp value to the min_temp / max_temp found
        in the underlying entity (if any)"""

        if not self.is_initialized:
            return value

        # Gets the min_temp and max_temp
        if (
            self._underlying_climate.min_temp is not None
            and self._underlying_climate is not None
        ):
            min_val = TemperatureConverter.convert(
                self._underlying_climate.min_temp, self._underlying_climate.temperature_unit, self._hass.config.units.temperature_unit
            )
            max_val = TemperatureConverter.convert(
                self._underlying_climate.max_temp, self._underlying_climate.temperature_unit, self._hass.config.units.temperature_unit
            )

            new_value = max(min_val, min(value, max_val))
        else:
            _LOGGER.debug("%s - no min and max attributes on underlying", self)
            new_value = value

        if new_value != value:
            _LOGGER.info(
                "%s - Target temp have been updated due min, max of the underlying entity. new_value=%.0f value=%.0f min=%.0f max=%.0f",
                self,
                new_value,
                value,
                min_val,
                max_val,
            )

        return new_value
```

L'algo prend le min et le max des entités sous-jacentes. Peux-tu regarder ce que tu as comme attributs dans tes entités sous-jacentes ? (Outils de dev / Etats)

**Comment by dliouville** [comment]
J'utilise la configuration centrale pour la PAC : 
![image](https://github.com/user-attachments/assets/23c73e83-263d-4740-945b-778598fe3fda)
et la configuration centrale définis : 
![image](https://github.com/user-attachments/assets/0aab94cc-34da-442c-989f-8c4a70341734)

**Comment by dliouville** [comment]
Mais effectivement, les thermostats par défaut de la PAC définissent des valeurs fantaisistes : 
![image](https://github.com/user-attachments/assets/ee763e18-90c1-441e-b302-5ff74df0a0d6)
Je ne vois pas d'ailleurs à quel endroit régler ces températures, je pensai que ce qui serait pris en compte ce sont les définitions de la configuration centrale.

**Comment by dliouville** [comment]
Ne voyant pas d'autre endroit où définir les valeurs min/max des sous jacents, je viens de le faire dans le yaml dev/etats. J'espère que ça sera bon du coup.

**Reply by jmcollin78** [reply]
> Ne voyant pas d'autre endroit où définir les valeurs min/max des sous jacents, je viens de le faire dans le yaml dev/etats. J'espère que ça sera bon du coup.

Non ca ne sera pas persistant. A chaque redémarrage tu devras le refaire.

Mais du coup, comme tes sous-jacents ont une plage qui va jusqu'à 30, ou est le problème ?
Je serais toi, je laisserai comme ça.
Tu as mis tes min et max dans la configuration centrale ce qui empeche de sélectionner plus ou moins par l'utilisateur mais la régulation interne peut aller plus loin. C'est très bien comme ça, ca permet d'utiliser à fond les capacités de ton sous-jacent. Ca peut surprendre, je suis d'accord, mais c'est finalement une très bonne chose.

**Reply by dliouville** [reply]
Le problème est que les modules de la PAC n'acceptent pas un réglage en dessous de 16°C, du coup comme je le disais, la régulation envoie des ordres à 12°C mais la PAC reste à 16°C ne pouvant pas aller en dessous et la régulation s'obstine à envoyer ses ordres inférieurs à 16°C, en boucle ... déclenchant un bip sonore à chaque fois sur mes unités intérieur de pac ... toutes les 5 minutes environ

**Comment by jmcollin78** [comment]
Y a un truc que je ne pige pas. Le clim_pac_salon_droite qui va de 8 à 30 c'est pas la PAC ?
![Capture d’écran 2024-10-27 à 17 41 04](https://github.com/user-attachments/assets/29ab7957-34f7-4bdb-929d-c0ac72030973)

**Reply by dliouville** [reply]
Oui, c'est une des unités de la PAC, mais elle n'est pas capable de réglage sur une plage de 8 à 30°C, sa vraie plage de réglage est de 16 à 30°C. Le "fabricant" Viessmann utilise des composants de la société chinoise Gree et les valeurs du contrôleur Gree ne correspondent pas aux spécifications finales du produit Viessmann.

**Reply by jmcollin78** [reply]
ok, j'ai compris. Désolé, j'ai été un peu long à la détente. Ce que tu vois depuis HA (8-30) n'est finalement pas accepté par la PAC. 

C'est embêtant de coupler le min/max du VTherm (ce que l'utilisateur peut changer) avec le min/max de la régulation (ce que le device accepte - ou est censé accepter dans ton cas). Je n'ai pas spécialement envie de pénaliser tout le monde pour ton cas très particulier.

Il te faudrait un climate virtuel qui pilote le climate pac_salon_droite. Ce climate virtuel aurait un bon min/max et serait celui qui est vu par VTherm. Ca fait un empilement de climate qui ne va pas aider au diagnostic en cas de soucis, mais ca devrait marcher.

Vtherm -> virtual climate (min=16) -> climate_pac_salon_droite

J'ai déjà fait des climate virtuels pour mes tests de la manière suivante:
```
climate:
    - platform: generic_thermostat
      name: Underlying thermostat1
      heater: input_boolean.fake_heater_switch3
      target_sensor: input_number.fake_temperature_sensor1
```

Regarde dans la doc (generic_thermostat) si on peut aller plus loin et spécifier des min / max et surtout commander le climate pac_salon. Je n'ai jamais fait ça.

**Reply by dliouville** [reply]
OK merci, je vais regarder dans ce sens là, j'imaginais sinon un script qui se lance à chaque redémarrage pour forcer la définition des valeurs.
